File: apps/reading/src/api/draftBeforeMergeCollectivePresentedTwinsCompose.ts
/**
 * Draft-before-full-merge gate over collective presented twins + paid ND pack
 * (pure).
 *
 * Operator vision: after multi-select collective of presented twins, create a
 * provisional combined draft before fully merging into the parent asset.
 * Full merge still requires separate full_merge_ack — never executes.
 *
 * draft_written / merge_executed / live_dispatched always false.
 * purchase_executed / live_router_authorized always false.
 * production_router_verdict always REJECT.
 */

import {
  composeFloatingDraftBeforeFullMergeGate,
  type FloatingDraftBeforeFullMergeGateCompose,
  type FloatingDraftBeforeFullMergeGateInput,
} from "./floatingDraftBeforeFullMergeGateCompose";
import {
  composeCollectivePresentedTwinsPaidNd,
  type CollectivePresentedTwinsPaidNdCompose,
  type CollectivePresentedTwinsPaidNdInput,
} from "./collectivePresentedTwinsPaidNdCompose";

export interface DraftBeforeMergeCollectivePresentedTwinsInput {
  draft_gate: Omit<FloatingDraftBeforeFullMergeGateInput, "operator_ack">;
  collective_pack: Omit<CollectivePresentedTwinsPaidNdInput, "operator_ack">;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface DraftBeforeMergeCollectivePresentedTwinsCompose {
  session_id: string;
  parent_asset_id: string;
  title: string;
  account_id: string;
  week_id: string;
  asset_id: string;
  draft_gate: FloatingDraftBeforeFullMergeGateCompose;
  collective_pack: CollectivePresentedTwinsPaidNdCompose;
  pack_ready: boolean;
  draft_written: false;
  merge_executed: false;
  live_dispatched: false;
  pack_dispatched: false;
  analysis_written: false;
  purchase_executed: false;
  charge_executed: false;
  hosted: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  live_router_authorized: false;
  twin_written: false;
  prompts_injected: false;
  live_dispatch_authorized: false;
  remote_fetched: false;
  backlog_mutated: false;
  secrets_stored: false;
  live_meter_read: false;
  store_mutated: false;
  suite_rewritten: false;
  live_execution_authorized: false;
  remote_index_queried: false;
  inventory_mutated: false;
  record_persisted: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "draft_before_merge_collective_presented_twins_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Draft-before-full-merge stacked on collective presented twins pack.
 * Never writes draft; never merges; never purchases; ND REJECT.
 */
export function composeDraftBeforeMergeCollectivePresentedTwins(
  input: DraftBeforeMergeCollectivePresentedTwinsInput,
): DraftBeforeMergeCollectivePresentedTwinsCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.draft_gate || typeof input.draft_gate !== "object") {
    throw new Error("draft_gate must be an object");
  }
  if (!input.collective_pack || typeof input.collective_pack !== "object") {
    throw new Error("collective_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "draft_written=false · merge_executed=false · live_dispatched=false",
    "purchase_executed=false · live_router_authorized=false",
    "production_router_verdict=REJECT",
  ];

  const draft_gate = composeFloatingDraftBeforeFullMergeGate({
    ...input.draft_gate,
    operator_ack: input.operator_ack,
  });
  notes.push(...draft_gate.notes.map((n) => `[draft_gate] ${n}`));

  const collective_pack = composeCollectivePresentedTwinsPaidNd({
    ...input.collective_pack,
    operator_ack: input.operator_ack,
  });
  notes.push(
    ...collective_pack.notes.map((n) => `[collective_pack] ${n}`),
  );

  const session_id = requireNonEmpty(draft_gate.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    draft_gate.parent_asset_id,
    "parent_asset_id",
  );
  const title = requireNonEmpty(collective_pack.title, "title");
  const account_id = requireNonEmpty(collective_pack.account_id, "account_id");
  const week_id = requireNonEmpty(collective_pack.week_id, "week_id");
  const asset_id = requireNonEmpty(collective_pack.asset_id, "asset_id");

  const session_aligned = collective_pack.session_id === session_id;
  const parent_aligned = collective_pack.parent_asset_id === parent_asset_id;
  if (!session_aligned) {
    notes.push(
      "session_id mismatch between draft_gate and collective_pack — pack_ready blocked",
    );
  }
  if (!parent_aligned) {
    notes.push(
      "parent_asset_id mismatch between draft_gate and collective_pack — pack_ready blocked",
    );
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      draft_gate.gate_ready === true &&
      collective_pack.pack_ready === true &&
      draft_gate.draft_written === false &&
      draft_gate.merge_executed === false &&
      draft_gate.live_dispatched === false &&
      collective_pack.live_dispatched === false &&
      collective_pack.merge_executed === false &&
      collective_pack.purchase_executed === false &&
      collective_pack.live_router_authorized === false &&
      collective_pack.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      input.operator_ack === true &&
      draft_gate.merge_executed === false &&
      collective_pack.purchase_executed === false &&
      collective_pack.production_router_verdict === "REJECT" &&
      (draft_gate.gate_ready === true || collective_pack.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — draft-before-merge + collective presented twins ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — draft_gate, collective_pack, alignment, or operator_ack gate open",
    );
  }

  if (
    draft_gate.draft_written !== false ||
    draft_gate.merge_executed !== false ||
    draft_gate.live_dispatched !== false ||
    collective_pack.live_dispatched !== false ||
    collective_pack.merge_executed !== false ||
    collective_pack.purchase_executed !== false ||
    collective_pack.live_router_authorized !== false ||
    collective_pack.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("draft_written=false");
  notes.push("merge_executed=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("analysis_written=false");
  notes.push("purchase_executed=false");
  notes.push("charge_executed=false");
  notes.push("hosted=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("live_router_authorized=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("remote_fetched=false");
  notes.push("backlog_mutated=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("store_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("live_execution_authorized=false");
  notes.push("remote_index_queried=false");
  notes.push("inventory_mutated=false");
  notes.push("record_persisted=false");
  notes.push("production_router_verdict=REJECT");

  return {
    session_id,
    parent_asset_id,
    title,
    account_id,
    week_id,
    asset_id,
    draft_gate,
    collective_pack,
    pack_ready,
    draft_written: false,
    merge_executed: false,
    live_dispatched: false,
    pack_dispatched: false,
    analysis_written: false,
    purchase_executed: false,
    charge_executed: false,
    hosted: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    live_router_authorized: false,
    twin_written: false,
    prompts_injected: false,
    live_dispatch_authorized: false,
    remote_fetched: false,
    backlog_mutated: false,
    secrets_stored: false,
    live_meter_read: false,
    store_mutated: false,
    suite_rewritten: false,
    live_execution_authorized: false,
    remote_index_queried: false,
    inventory_mutated: false,
    record_persisted: false,
    production_router_verdict: "REJECT",
    notes,
    authority: "draft_before_merge_collective_presented_twins_compose_advisory",
  };
}

export function formatDraftBeforeMergeCollectivePresentedTwinsSummary(
  c: DraftBeforeMergeCollectivePresentedTwinsCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `gate_ready=${c.draft_gate.gate_ready} · ` +
    `stage=${c.draft_gate.stage} · ` +
    `collective_ready=${c.collective_pack.pack_ready} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `draft_written=false · merge_executed=false · purchase_executed=false`
  );
}

File: substrate/draft_before_merge_collective_presented_twins_compose.py
"""Draft-before-full-merge over collective presented twins + paid ND pack.

draft_written / merge_executed / live_dispatched always False.
purchase_executed / live_router_authorized always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.collective_presented_twins_paid_nd_compose import (
    CollectivePresentedTwinsPaidNdCompose,
    CollectivePresentedTwinsPaidNdComposeError,
    compose_collective_presented_twins_paid_nd,
)
from substrate.floating_draft_before_full_merge_gate_compose import (
    FloatingDraftBeforeFullMergeGateCompose,
    FloatingDraftBeforeFullMergeGateComposeError,
    compose_floating_draft_before_full_merge_gate,
)


class DraftBeforeMergeCollectivePresentedTwinsComposeError(ValueError):
    """Fail-closed validation for draft-before-merge + collective pack."""


@dataclass(frozen=True)
class DraftBeforeMergeCollectivePresentedTwinsCompose:
    session_id: str
    parent_asset_id: str
    title: str
    account_id: str
    week_id: str
    asset_id: str
    draft_gate: FloatingDraftBeforeFullMergeGateCompose
    collective_pack: CollectivePresentedTwinsPaidNdCompose
    pack_ready: bool
    draft_written: bool
    merge_executed: bool
    live_dispatched: bool
    pack_dispatched: bool
    analysis_written: bool
    purchase_executed: bool
    charge_executed: bool
    hosted: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    live_router_authorized: bool
    twin_written: bool
    prompts_injected: bool
    live_dispatch_authorized: bool
    remote_fetched: bool
    backlog_mutated: bool
    secrets_stored: bool
    live_meter_read: bool
    store_mutated: bool
    suite_rewritten: bool
    live_execution_authorized: bool
    remote_index_queried: bool
    inventory_mutated: bool
    record_persisted: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "draft_gate": self.draft_gate.to_dict(),
            "collective_pack": self.collective_pack.to_dict(),
            "pack_ready": self.pack_ready,
            "draft_written": False,
            "merge_executed": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "analysis_written": False,
            "purchase_executed": False,
            "charge_executed": False,
            "hosted": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "live_router_authorized": False,
            "twin_written": False,
            "prompts_injected": False,
            "live_dispatch_authorized": False,
            "remote_fetched": False,
            "backlog_mutated": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "store_mutated": False,
            "suite_rewritten": False,
            "live_execution_authorized": False,
            "remote_index_queried": False,
            "inventory_mutated": False,
            "record_persisted": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "draft_before_merge_collective_presented_twins_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_draft_before_merge_collective_presented_twins(
    *,
    draft_gate: object,
    collective_pack: object,
    operator_ack: object,
    require_both: object | None = None,
) -> DraftBeforeMergeCollectivePresentedTwinsCompose:
    """Draft-before-merge + collective pack. Never writes or merges."""
    if not isinstance(operator_ack, bool):
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(draft_gate, dict):
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            "draft_gate must be an object"
        )
    if not isinstance(collective_pack, dict):
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            "collective_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "draft_written=false · merge_executed=false · live_dispatched=false",
        "purchase_executed=false · live_router_authorized=false",
        "production_router_verdict=REJECT",
    ]

    try:
        dg = compose_floating_draft_before_full_merge_gate(
            session_id=draft_gate.get("session_id"),
            parent_asset_id=draft_gate.get("parent_asset_id"),
            sources=draft_gate.get("sources"),
            stage=draft_gate.get("stage"),
            operator_ack=operator_ack,
            parent_excerpt=draft_gate.get("parent_excerpt"),
            full_merge_ack=draft_gate.get("full_merge_ack"),
        )
    except FloatingDraftBeforeFullMergeGateComposeError as e:
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            str(e)
        ) from e
    notes.extend(f"[draft_gate] {n}" for n in dg.notes)

    try:
        col = compose_collective_presented_twins_paid_nd(
            collective=collective_pack.get("collective"),
            paid_nd=collective_pack.get("paid_nd"),
            operator_ack=operator_ack,
            require_both=collective_pack.get("require_both"),
        )
    except CollectivePresentedTwinsPaidNdComposeError as e:
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            str(e)
        ) from e
    notes.extend(f"[collective_pack] {n}" for n in col.notes)

    session = _require_nonempty(dg.session_id, field="session_id")
    parent = _require_nonempty(dg.parent_asset_id, field="parent_asset_id")
    title = _require_nonempty(col.title, field="title")
    account = _require_nonempty(col.account_id, field="account_id")
    week = _require_nonempty(col.week_id, field="week_id")
    asset = _require_nonempty(col.asset_id, field="asset_id")

    session_aligned = col.session_id == session
    parent_aligned = col.parent_asset_id == parent
    if not session_aligned:
        notes.append(
            "session_id mismatch between draft_gate and collective_pack — "
            "pack_ready blocked"
        )
    if not parent_aligned:
        notes.append(
            "parent_asset_id mismatch between draft_gate and collective_pack — "
            "pack_ready blocked"
        )

    if require:
        pack_ready = (
            session_aligned
            and parent_aligned
            and dg.gate_ready is True
            and col.pack_ready is True
            and dg.draft_written is False
            and dg.merge_executed is False
            and dg.live_dispatched is False
            and col.live_dispatched is False
            and col.merge_executed is False
            and col.purchase_executed is False
            and col.live_router_authorized is False
            and col.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            session_aligned
            and parent_aligned
            and operator_ack is True
            and dg.merge_executed is False
            and col.purchase_executed is False
            and col.production_router_verdict == "REJECT"
            and (dg.gate_ready is True or col.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — draft-before-merge + collective presented twins "
            "ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — draft_gate, collective_pack, alignment, or "
            "operator_ack gate open"
        )

    if (
        dg.draft_written is not False
        or dg.merge_executed is not False
        or dg.live_dispatched is not False
        or col.live_dispatched is not False
        or col.merge_executed is not False
        or col.purchase_executed is not False
        or col.live_router_authorized is not False
        or col.production_router_verdict != "REJECT"
    ):
        raise DraftBeforeMergeCollectivePresentedTwinsComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "draft_written=false",
            "merge_executed=false",
            "live_dispatched=false",
            "pack_dispatched=false",
            "analysis_written=false",
            "purchase_executed=false",
            "charge_executed=false",
            "hosted=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "live_router_authorized=false",
            "twin_written=false",
            "prompts_injected=false",
            "live_dispatch_authorized=false",
            "remote_fetched=false",
            "backlog_mutated=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "store_mutated=false",
            "suite_rewritten=false",
            "live_execution_authorized=false",
            "remote_index_queried=false",
            "inventory_mutated=false",
            "record_persisted=false",
            "production_router_verdict=REJECT",
        )
    )

    return DraftBeforeMergeCollectivePresentedTwinsCompose(
        session_id=session,
        parent_asset_id=parent,
        title=title,
        account_id=account,
        week_id=week,
        asset_id=asset,
        draft_gate=dg,
        collective_pack=col,
        pack_ready=pack_ready,
        draft_written=False,
        merge_executed=False,
        live_dispatched=False,
        pack_dispatched=False,
        analysis_written=False,
        purchase_executed=False,
        charge_executed=False,
        hosted=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        live_router_authorized=False,
        twin_written=False,
        prompts_injected=False,
        live_dispatch_authorized=False,
        remote_fetched=False,
        backlog_mutated=False,
        secrets_stored=False,
        live_meter_read=False,
        store_mutated=False,
        suite_rewritten=False,
        live_execution_authorized=False,
        remote_index_queried=False,
        inventory_mutated=False,
        record_persisted=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority=(
            "draft_before_merge_collective_presented_twins_compose_advisory"
        ),
    )


def format_draft_before_merge_collective_presented_twins_summary(
    c: DraftBeforeMergeCollectivePresentedTwinsCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"gate_ready={c.draft_gate.gate_ready} · "
        f"stage={c.draft_gate.stage} · "
        f"collective_ready={c.collective_pack.pack_ready} · "
        f"verdict={c.production_router_verdict} · "
        f"draft_written=false · merge_executed=false · purchase_executed=false"
    )


__all__ = [
    "DraftBeforeMergeCollectivePresentedTwinsCompose",
    "DraftBeforeMergeCollectivePresentedTwinsComposeError",
    "compose_draft_before_merge_collective_presented_twins",
    "format_draft_before_merge_collective_presented_twins_summary",
]
